### PR TITLE
[C#] Add cancellation token overloads to streaming interfaces

### PR DESCRIPTION
**Pull request by JamesNK**
See https://github.com/grpc/grpc-dotnet/issues/1422#issuecomment-943771494

* Adds default interface implementation with cancellation token overloads
* Default implementation calls the current interface method. grpc-dotnet will override default implementation to use the cancellation token
* Feature requires C# 8.0
* Feature isn't supported on .NET Framework. Overloads will only be available with .NET Standard 2.1 target (i.e. .NET Core 3 or later)

@jtattermusch @captainsafia

### CODE REVIEW

**Comment by JamesNK** [comment]
Error related to updating SDK:
```
2021-10-31 14:27:53,001 START: run_tests_csharp_linux_dbg_native
2021-10-31 14:27:53,006 START: run_tests_csharp_linux_opt_native
2021-10-31 14:27:54,732 ++ dirname tools/run_tests/dockerize/build_docker_and_run_tests.sh
+ cd tools/run_tests/dockerize/../../..
++ pwd
+ git_root=/tmpfs/src/github/grpc
+ cd -
/tmpfs/src/github/grpc
++ basename tools/dockerfile/test/csharp_buster_x64
++ sha1sum tools/dockerfile/test/csharp_buster_x64/Dockerfile
++ cut -f1 '-d '
+ DOCKER_IMAGE_NAME=csharp_buster_x64:89ca6f76aece9e346f28331ebd6deda8fd5b1220
+ '[' grpctesting '!=' '' ']'
+ DOCKER_IMAGE_NAME=grpctesting/csharp_buster_x64:89ca6f76aece9e346f28331ebd6deda8fd5b1220
+ docker pull grpctesting/csharp_buster_x64:89ca6f76aece9e346f28331ebd6deda8fd5b1220
Error response from daemon: manifest for grpctesting/csharp_buster_x64:89ca6f76aece9e346f28331ebd6deda8fd5b1220 not found
Traceback (most recent call last):
  File "tools/run_tests/run_tests.py", line 1607, in <module>
    env=env)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'tools/run_tests/dockerize/build_docker_and_run_tests.sh' returned non-zero exit status 1
```

**Review by captainsafia — APPROVED** [review]

**Comment by jtattermusch** [comment]
Unfortunately bumping the .NET SDK for Grpc.Core builds isn't that easy (more changes is needed in other places as well). I'd have to look into that separately.

**Comment by jtattermusch** [comment]
Note to self: Recent PR for upgrading the dotnet SDK: https://github.com/grpc/grpc/pull/26465/files

**Comment by jtattermusch** [comment]
I will look at this in detail once I successfully upgrade the dotnet SDK in the grpc/grpc repo: https://github.com/grpc/grpc/pull/27966 (it's proving difficult).

**Comment by jtattermusch** [comment]
SDK has been updated in #27966, you can rebase now.

**Comment by JamesNK** [comment]
ping @jtattermusch

File: src/csharp/Grpc.Core.Api/IClientStreamWriter.cs
Comment (jtattermusch) [inline src/csharp/Grpc.Core.Api/IClientStreamWriter.cs:46]: same concerns/questions as for WriteAsync

**Review by jtattermusch — COMMENTED** [review]
I wrote a few comments earlier, but forgot to hit "send", sorry for the delay.
See comments for concerns (To me it's unclear what the implementation would look like for Grpc.Core, and simply ignoring the cancellation token seems problematic).

Btw, in https://github.com/grpc/grpc-dotnet/issues/1422#issuecomment-961851240 I pointed out that we are potentially missing a mechanism to cancel the entire call on the server side (on the client side cancelling a call is easy since you can pass cancellation token to a new call). One can "cancel" the server side call handler by simple throwing/returning from the handler function, but there is no other way.

Exposing overloads that take cancellationToken on individual write operations is basically just a proxy for cancelling the entire call, since the semantics of "cancel an individual write operation" is non-recoverable and the entire call will need to be cancelled anyway.

So perhaps we should first figure out the mechanism for cancelling the entire server side call (and expose it e.g. on ServerCallContext) after which adding the new WriteAsync overloads that take cancellationToken would become just syntactic sugar?

File: src/csharp/Grpc.Core.Api/IClientStreamWriter.cs
Comment (jtattermusch) [inline src/csharp/Grpc.Core.Api/IClientStreamWriter.cs:44]: Also this overload only exists for client-side call (since this is IClientStreamWriter) and on writer we already have a simple mechanism for cancelling a call (you pass a cancellation token to a newly created call), so adding `CompleteAsync(CancellationToken cancellationToken)` overload is not going to help with the situation on the server and it's redundant on the client.

File: src/csharp/Grpc.Core.Api/IAsyncStreamWriter.cs
Comment (JamesNK) [reply src/csharp/Grpc.Core.Api/IAsyncStreamWriter.cs:46]: Kestrel supports aborting in-progress requests. I haven't looked into exactly how it will work but it will either be:
1. The cancellation token is passed to the underlying writer and if the writer gets a cancellation then it will abort the request, or
2. WriteAsync subscribes to the cancellation token with a callback that calls `HttpContext.Abort()`.

I don't know the internals of Grpc.Core.

File: src/csharp/Grpc.Core.Api/IClientStreamWriter.cs
Comment (JamesNK) [reply src/csharp/Grpc.Core.Api/IClientStreamWriter.cs:44]: Makes sense. I'll remove it.

File: src/csharp/Grpc.Core.Api/IAsyncStreamWriter.cs
Comment (JamesNK) [reply src/csharp/Grpc.Core.Api/IAsyncStreamWriter.cs:46]: > Grpc.Core

Would Grpc.Core cancellation be supported with this code:

https://github.com/grpc/grpc/blob/84101427d0a699e65837af43f4f0ac326c7a2aca/src/csharp/Grpc.Core/Internal/AsyncCallBase.cs#L393-L399

It is used by `IAsyncStreamReader<T>.MoveNext` which takes a cancellation token:

https://github.com/grpc/grpc/blob/2d4f3c56001cd1e1f85734b2f7c5ce5f2797c38a/src/csharp/Grpc.Core/Internal/ServerRequestStream.cs#L50-L58

Comment (JamesNK) [reply src/csharp/Grpc.Core.Api/IAsyncStreamWriter.cs:46]: > Grpc.Core

Actually, Grpc.Core only supports up to netstandard20 and net45. The new overload won't be available.

**Review by jtattermusch — APPROVED** [review]
LGTM.

**Comment by jtattermusch** [comment]
@apolcyn  please add another LGTM.

**Comment by jtattermusch** [comment]
Added "RELEASE BLOCKER" to make sure it gets included in the upcoming 1.45.x branch cut.
The PR is ready to merge once @apolcyn  approves.

**Review by apolcyn — APPROVED** [review]

**Pull request merged**
